be/src/olap/schema_change.cpp

```diff
@@ -1121,6 +1121,207 @@ bool SchemaChangeWithSorting::_external_sorting(
+              << ", alter_version_hash=" << request.alter_version_hash;
+
+    // Lock schema_change_lock util schema change info is stored in tablet header
+    if (!StorageEngine::instance()->tablet_manager()->try_schema_change_lock(request.base_tablet_id)) {
+        LOG(WARNING) << "failed to obtain schema change lock. "
+                     << "base_tablet=" << request.base_tablet_id;
+        return OLAP_ERR_TRY_LOCK_FAILED;
+    }
+
+    OLAPStatus res = _do_process_alter_tablet_v2(request);
+    LOG(INFO) << "finished alter tablet process, res=" << res;
+    StorageEngine::instance()->tablet_manager()->release_schema_change_lock(request.base_tablet_id);
+    return res;
+}
+
+OLAPStatus SchemaChangeHandler::_do_process_alter_tablet_v2(const TAlterTabletReqV2& request) {
+    OLAPStatus res = OLAP_SUCCESS;
+    TabletSharedPtr base_tablet = StorageEngine::instance()->tablet_manager()->get_tablet(
+            request.base_tablet_id, request.base_schema_hash);
+    if (base_tablet == nullptr) {
+        LOG(WARNING) << "fail to find base tablet. base_tablet=" << request.base_tablet_id
+                     << ", base_schema_hash=" << request.base_schema_hash;
+        return OLAP_ERR_TABLE_NOT_FOUND;
+    }
+
+    // new tablet has to exist
+    TabletSharedPtr new_tablet = StorageEngine::instance()->tablet_manager()->get_tablet(
+            request.new_tablet_id, request.new_schema_hash);
+    if (new_tablet == nullptr) {
+        LOG(WARNING) << "fail to find new tablet."
+                     << ", new_tablet=" << request.new_tablet_id
+                     << ", new_schema_hash=" << request.new_schema_hash;
+        return OLAP_ERR_TABLE_NOT_FOUND;
+    }
+
+    // check if tablet's state is not_ready, if it is ready, it means the tablet already finished 
+    // check whether the tablet's max continuous version == request.version
+    if (new_tablet->tablet_state() != TABLET_NOTREADY) {
+        res = _validate_alter_result(new_tablet, request);
+        LOG(INFO) << "tablet's state=" << new_tablet->tablet_state()
+                  << " the convert job alreay finished, check its version"
+                  << " res=" << res;
+        return res;
+    }
+
+    LOG(INFO) << "finish to validate alter tablet request. begin to convert data from base tablet to new tablet" 
+              << " base_tablet=" << base_tablet->full_name()
+              << " new_tablet=" << new_tablet->full_name();
+
+    ReadLock base_migration_rlock(base_tablet->get_migration_lock_ptr(), TRY_LOCK);
+    if (!base_migration_rlock.own_lock()) {
+        return OLAP_ERR_RWLOCK_ERROR;
+    }
+    ReadLock new_migration_rlock(new_tablet->get_migration_lock_ptr(), TRY_LOCK);
+    if (!new_migration_rlock.own_lock()) {
+        return OLAP_ERR_RWLOCK_ERROR;
+    }
+
+    // 2. Get version_to_be_changed and store into tablet header
+    base_tablet->obtain_push_lock();
+    base_tablet->obtain_header_wrlock();
+    new_tablet->obtain_header_wrlock();
```

Comment:
Why not requiring new_tablet's push lock？
I think you should add comment to explain which locks should be held at the beginning of alter job method. Or the code is very difficult to maintain.

Reply:
Done